### ISSUE TITLE
ENH: broadcasting perf on C-contiguous arrays

### ISSUE THREAD

**Issue by tylerjereddy** [numpy/numpy#13307]
This came up in a recent community call, as a point of discussion on behalf of @zemanj. We didn't discuss it much in their absence, but I am placing the same summary of the idea here for potential interest/ assessment. Mostly for the low-level gurus, I suspect.

> The way NumPy implements broadcasting when executing operations such as `matrix += vector` on C-contiguous arrays apparently copies data to intermediate buffers more often than necessary before calling vectorized routines (e.g., `sse2_binary_add_FLOAT()`). See also the `gdb` log in [this comment](https://github.com/MDAnalysis/mdanalysis/issues/2210#issuecomment-480805965) and the `kcachegrind` visualization in [this comment](https://github.com/MDAnalysis/mdanalysis/issues/2210#issuecomment-480817042). There is [some interest](https://github.com/MDAnalysis/mdanalysis/pull/2211#discussion_r274025811) in shortcuts around `npyiter_copy_to_buffers()`

**Comment by zemanj** [comment]
Thank you @tylerjereddy for opening this issue! I'll go ahead and try to provide some more details.

I'm looking foreward to your questions / comments / suggestions!

### Scenario demonstrating the problem

Consider the simple case where we have a 2d ndarray `matrix` of shape, say, `(100000, 3)` and dtype `float32` (could probably also be any other dtype), and another 1d ndarray `vector` of shape `(3,)` and the same dtype as `matrix`.
What we now want to do is translate every row in `matrix` by `vector`, so we execute
```Python
matrix += vector
```
First, let's assume that `matrix` is Fortran-contiguous and measure the run time of the in-place addition:
```Python
import numpy as np

matrix = np.ones((100000, 3), dtype=np.float32, order='F')
vector = np.zeros((3,), dtype=np.float32)   # use zeros to prevent overflow during timing

def inplace_translate(mat, vec):
    mat += vec

%timeit inplace_translate(matrix, vector)
```
This results in
```
26.4 µs ± 391 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
Now we change the memory layout of `matrix` to C-contiguous and see how the same in-place addition performs:
```Python
import numpy as np

matrix = np.ones((100000, 3), dtype=np.float32, order='C')  # order is the only difference!
vector = np.zeros((3,), dtype=np.float32)

def inplace_translate(mat, vec):
    mat += vec

%timeit inplace_translate(matrix, vector)
```
Assuming a naive worst-case implementation looping over the `matrix` columns in the inner loop, we'd expect the operation to take roughly three times as long as before since the array has to be fetched from memory three times. The shocking (!) result, however, is this:
```
420 µs ± 1.59 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
Changing the memory layout of `matrix` to C-contiguous has increased the run time by a factor of **15.9**!
Obviously, there must be something fishy going on in the background.

### Tracing the root of all evil
Running the above code in `gdb` shows that numpy correctly iterates over the `matrix` array row-wise so that memory access is not strided, and both `matrix` and `vector` are fetched from memory only once. So that's not causing the problem.

What `gdb` also reveils is that the addition is performed using SIMD-accelerated vector operations as implemented [here](https://github.com/numpy/numpy/blob/4a3396322d7f73d77d4d229c78410e7b808738a5/numpy/core/src/umath/simd.inc.src#L428), which is in principle a good thing. To this aim, since the data pointer of the `vector` array cannot be generally assumed to be aligned to `xmm/ymm/zmm` register size (depending on the type of SIMD acceleration, i.e., SSE2/AVX(2)/AVX512, respectively), the data contained in `vector` is `memmove`d to a correctly aligned buffer. Since we are dealing with an in-place operation here, the data from `matrix` is not moved to an intermediate buffer but fetched with unaligned loads (`__m128_loadu_ps` in case of SSE2) instead. On modern CPUs, the performance penalty of unaligned loads/stores is usually negligible, so this should be nothing to worry about, and it's definitely not slower than copying the whole `matrix` array to an aligned buffer anyway. Up to here, this is all fine.

What's really causing the slow-down is how broadcasting of the `vector` data onto each row of `matrix` is handled in [`iterator_loop()`](https://github.com/numpy/numpy/blob/e8689171bb5830e7c607d4c398353dc8707c0106/numpy/core/src/umath/ufunc_object.c#L1461). First, numpy allocates a correctly aligned 32 KiB buffer, which is then filled with copies of `vector`. In the above scenario, this means that `ceil(32768 // 12) == 2731` `memmove()` operations are necessary to fill that buffer once. Then, the vectorized additions of the first 32 KiB of `matrix` and that buffer are performed. This process is then performed for each of the remaining 32 KiB chunks of `matrix`, while the buffer is refilled with corresponding copies of `vector` **in each and every iteration**. Thus, the above scenario requires more than 100,000 `memmove`s! It is quite obvious that this will be... well... rather slow.

### Expected behavior

* The minimum required buffer size is the lowest common multiple of `vector.itemsize * vector.size` and the SIMD register size (unless that exceeds a certain maximum size). Optimally, it is also a multiple of the CPU's cache line size (usually 64 bytes) to prevent false sharing in operations implementing multi-threaded buffer writes. In the above scenario, that would result in a 192 bytes buffer (16 copies of `vector` resulting in three 64-byte cache lines or 12 SSE2 (6 AVX / 3 AVX512) register sizes).

* The maximum allowed buffer size should be chosen so that it fits in the CPU's L1d cache *together* with a correspondingly sized chunk of the other operand's data and, for out-of-place operations, the result buffer. Since a very common L1d cache size is 32 KiB, the current choice is suboptimal, since the buffer can never be entirely held in L1d.

* The buffer should be filled with contents of `vector` exactly *once* and reused in successive operations. For the above scenario, this means that only 16 instead of more than 100,000 `memmove`s are required.

**Comment by zemanj** [comment]
## Cache efficiency analysis

**tl;dr**: The relative number of cache misses are *not* a problem. Thus, against my first intuition, the buffer size of 32KiB is not that problematic (seems like the prefetcher is usually smart enough). The slow-down for `order='C'` is mainly due to the additional CPU cycles spent moving data around in cache.

If you want to know some details, here they are:

### Measurements
Test CPU:
```
Model name:    Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
L1d cache:     32K
L1i cache:     32K
L2 cache:      256K
L3 cache:      8192K
```
I used the following bash script to check for possible cache misses with `perf`:
```bash
#! /bin/bash

# record these perf events:
events=cpu-cycles
events=${events},cache-references
events=${events},cache-misses
events=${events},LLC-loads
events=${events},LLC-load-misses
events=${events},LLC-stores
events=${events},LLC-store-misses
events=${events},L1-dcache-load-misses
events=${events},L1-dcache-loads
events=${events},L1-dcache-stores

# pin process to core 0:
cpu_id=0

# wait one second before recording perf counters
delay=1000

# choose memory layout:
order=F  # or order=F

#Python code to execute
read -r -d '' pycode << EOF
import numpy as np
matrix = np.ones((100000, 3), dtype=np.float32, order='$order')
vector = np.zeros((3,), dtype=np.float32)
for i in range(100000):
    matrix += vector
EOF

# print overall perf statistics
perf stat -D $delay -C $cpu_id -e $events \
numactl --physcpubind=$cpu_id --localalloc python3 -c """$pycode"""

# record detailed perf statistics
numactl --physcpubind=$cpu_id --localalloc python3 -c """$pycode""" & pid=$!
perf record -D $delay -p $pid -e $events --
```
### Results for order='F'
#### perf stat output
```
Performance counter stats for 'CPU(s) 0':

    9,868,979,194      cpu-cycles                                                    (39.97%)
    2,923,453,554      cache-references                                              (40.08%)
        2,124,407      cache-misses              #    0.073 % of all cache refs      (40.08%)
      300,894,404      LLC-loads                                                     (40.08%)
           61,857      LLC-load-misses           #    0.02% of all LL-cache hits     (40.06%)
        2,901,490      LLC-stores                                                    (19.97%)
              145      LLC-store-misses                                              (19.97%)
    1,384,952,760      L1-dcache-load-misses     #   24.38% of all L1-dcache hits    (29.96%)
    5,680,989,724      L1-dcache-loads                                               (39.95%)
    5,571,802,127      L1-dcache-stores                                              (39.95%)

      3.202966799 seconds time elapsed
```
#### selected perf record results
```
cpu-cycles:
  90.95%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
cache-references:
  93.34%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
cache-misses:
  [about 60% python3 kernel functions]
  5.39%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
LLC-loads:
  95.05%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
LLC-load-misses:
  48.21%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
  [rest are mostly python3 kernel calls]
L1-dcache-loads:
  94.50%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
L1-dcache-load-misses:
  97.97%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
```

### Results for order='C'
#### perf stat output
```
Performance counter stats for 'CPU(s) 0':

   140,306,210,106      cpu-cycles                                                    (40.00%)
     4,018,541,682      cache-references                                              (40.00%)
         2,997,337      cache-misses              #    0.075 % of all cache refs      (40.01%)
        92,168,208      LLC-loads                                                     (40.00%)
           163,644      LLC-load-misses           #    0.18% of all LL-cache hits     (40.00%)
         4,443,652      LLC-stores                                                    (20.00%)
            39,424      LLC-store-misses                                              (20.00%)
     5,443,110,974      L1-dcache-load-misses     #    4.11% of all L1-dcache hits    (30.00%)
   132,518,690,472      L1-dcache-loads                                               (40.00%)
    46,940,829,776      L1-dcache-stores                                              (40.00%)

      32.322642571 seconds time elapsed
```
#### selected perf record results
```
cpu-cycles:
  40.87%  python3  libc-2.27.so                                       [.] __memmove_avx_unaligned_erms
  31.03%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] PyArray_TransferNDimToStrided
  12.73%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] _contig_to_contig
  12.33%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
cache-references:
  91.51%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
cache-misses:
  22.48%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
  18.44%  python3  libc-2.27.so                                       [.] __memmove_avx_unaligned_erms
  10.19%  python3  [kernel]
   8.97%  python3  libc-2.27.so                                       [.] cfree@GLIBC_2.2.5
   8.42%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] _contig_to_contig
   3.46%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] PyArray_TransferNDimToStrided
LLC-loads:
  88.47%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
LLC-load-misses:
  32.38%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
  [rest are mostly python3 kernel calls]
L1-dcache-loads:
  39.79%  python3  libc-2.27.so                                       [.] __memmove_avx_unaligned_erms
  36.44%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] PyArray_TransferNDimToStrided
  10.92%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
  10.76%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] _contig_to_contig
L1-dcache-load-misses:
  62.74%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] FLOAT_add
  20.69%  python3  libc-2.27.so                                       [.] __memmove_avx_unaligned_erms
  12.31%  python3  _multiarray_umath.cpython-36m-x86_64-linux-gnu.so  [.] PyArray_TransferNDimToStrided
```

**Comment by seberg** [comment]
Marking as project: this needs deep digging into relatively complex C code.  But I can give pointers to understand the weirdness of the logic and its doable (I have fixed a deep bug here when new to the code ;)).

My main reason for marking: looking over, this, I do think the OP is right.  There are some very low hanging fruits for big speed improvements by tweaking the buffer sizes.

Even better, I wonder if we can tweak the buffer sizes aggressively enough that we can simplify the buffering logic _a lot_.
